syntax="proto2";

package sogou.nlu.rpc;

option java_package = "sogou.nlu.rpc";
option java_outer_classname = "sogou.nlu.rpc";
option py_generic_services = true;
option java_generic_services = true;
option cc_generic_services = true;


message Request {

  // RPC service full name
  required string service_name = 1;
  
  // RPC method name
  required string method_name = 2;
  
  // RPC request proto
  required bytes request_proto = 3;
}


message Response {

  // Error Reason
  required NErrorCode error_code = 1;

  // Error, if any
  optional string error = 2;
  
  // RPC response proto
  required bytes response_proto = 3;
}


// Possible error reasons
// The server-side errors are returned in the response from the server.
// The client-side errors are returned by the client-side code when it doesn't 
// have a response from the server.
enum NErrorCode {

  NOERROR = 0; // No error

  // Server-side errors
  BAD_REQUEST_DATA = 1; // Server received bad request data
  BAD_REQUEST_PROTO = 2; // Server received bad request proto
  SERVICE_NOT_FOUND = 3; // Service not found on server
  METHOD_NOT_FOUND = 4; // Method not found on server
  RPC_ERROR = 5; // Rpc threw exception on server
  RPC_FAILED = 6; // Rpc failed on server
  
  // Client-side errors (these are returned by the client-side code)
  INVALID_REQUEST_PROTO = 7; // Rpc was called with invalid request proto
  BAD_RESPONSE_PROTO = 8; // Server returned a bad response proto
  UNKNOWN_HOST = 9; // Could not find supplied host
  IO_ERROR = 10; // I/O error while communicating with server
}
